warmup.py

```diff
@@ -103,7 +105,23 @@ def energy(self, r):
             If ``r`` is 0, the energy is :py:obj:`numpy.inf`.
 
         """
-        pass
+        if isinstance(r, (list, tuple, np.ndarray)):
+            u = []
+            for ri in r:
+                if ri == 0:
```

Comment:
It is not usually a good idea to test if a floating-point value is exactly equal to another value (although 0 is OK because it is exactly representable). Instead, you should use `np.isclose`.